### PR TITLE
Add test for replica auto balance

### PR DESCRIPTION
**Pull request by c3y1huang**
https://github.com/longhorn/longhorn/issues/587

### CODE REVIEW

**Comment by c3y1huang** [comment]
Regression * 2 - [Build#259](https://ci.longhorn.io/job/private/job/longhorn-tests-regression/259/testReport/tests/)

File: manager/integration/tests/test_zone.py [hunk 1]
```diff
@@ -301,3 +330,611 @@ def test_replica_zone_anti_affinity(client, core_api, volume_name, k8s_node_zone
...
+        set node-3 to zone-2.
+    And create volume with 2 replicas.
+    And create pv for volume.
+    And create pvc for volume.
+
+    Scenario: replica auto-balance zones with best-effort should not
+              remove pod local replicas when data locality is enabled
+              (best-effort).
+
+    Given create and wait pod on <node>.
```
Comment (shuo-wu) [inline manager/integration/tests/test_zone.py:572]: `<node>`is confusing. You can specify the node name for a pod

File: manager/integration/tests/test_zone.py [hunk 2]
```diff
@@ -301,3 +330,611 @@ def test_replica_zone_anti_affinity(client, core_api, volume_name, k8s_node_zone
...
+    And 1 replica running in zone-1.
+        1 replica running in zone-2.
+
+    When sleep 10 seconds, to ensure no addition scheduling is happening.
+    Then count replicas on each nodes.
+    And 1 replica running on <pod-node>.
+        0 replica running on <duplicate-node>.
+        1 replica running on node-3.
+
+    And delete pod.
```
Comment (shuo-wu) [inline manager/integration/tests/test_zone.py:594]: You can launch the pod on another node of zone1 then see if the replica in zone1 will be migrated to the new node.

**Review by shuo-wu — CHANGES_REQUESTED** [review]

File: manager/integration/tests/common.py [hunk 3]
```diff
@@ -1587,6 +1599,31 @@ def wait_for_volume_replicas_mode(client, volname, mode,
...
+
+            if replica_balanced:
+                hosts.remove(replica.hostId)
+
+            num_running += 1
+        if num_running == volume.numberOfReplicas:
+            break
+
+        time.sleep(RETRY_INTERVAL)
+    return num_running
```
[the remark this answers is not in the document]
Comment (c3y1huang) [reply manager/integration/tests/common.py]: Addressed.

File: manager/integration/tests/test_scheduling.py [hunk 4]
```diff
@@ -422,6 +424,376 @@ def test_replica_rebuild_per_volume_limit(
...
+
+    volume = create_and_check_volume(client, volume_name)
+    volume.attach(hostId=n1.name)
+    volume = wait_for_volume_healthy(client, volume_name)
+
+    data = write_volume_random_data(volume)
+    check_volume_data(volume, data)
+
+    client.update(n2, allowScheduling=False, evictionRequested=True)
+    client.update(n3, allowScheduling=False, evictionRequested=True)
```
[the remark this answers is not in the document]
Comment (c3y1huang) [reply manager/integration/tests/test_scheduling.py]: Thanks, fixed.

File: manager/integration/tests/test_scheduling.py [hunk 5]
```diff
@@ -422,6 +424,376 @@ def test_replica_rebuild_per_volume_limit(
...
+    And set `replica-auto-balance` to `least_effort`.
+    And create a volume.
+    And attach the volume to node-1.
+    And wait for the volume to be healthy.
+    And write some data to the volume.
+    And disable scheduling for node-2.
+        disable scheduling for node-3.
+    And set volume replicas to 6.
+    And count replicas running on each nodes.
+    And 6 replicas running on node-1.
```
[the remark this answers is not in the document]
Comment (c3y1huang) [reply manager/integration/tests/test_scheduling.py:441]: This is part of the setup in `Given` block, if possible I like to keep the way it is so the testing focus is clear (`When`, `Then`).

File: manager/integration/tests/test_scheduling.py [hunk 6]
```diff
@@ -422,6 +424,376 @@ def test_replica_rebuild_per_volume_limit(
...
+
+    When enable scheduling for node-2.
+         enable scheduling for node-3.
+    And count replicas running on each nodes for volume-1.
+        count replicas running on each nodes for volume-2.
+
+    Then volume-1 replicas should be running on node-1.
+    And volume-1 should have 3 replicas running.
+    And volume-2 replicas should be running on node-1, node-2, node-3.
+    And volume-2 should have 3 replicas running.
```
[the remark this answers is not in the document]
Comment (c3y1huang) [reply manager/integration/tests/test_scheduling.py]: > When enable scheduling for node-2 and node-3`

I like to keep the way it is if possible, IMO it's easier to read when tracking the test code. 

> Then all 3 replicas of volume-1 should be running on node-1
> And all 3 replicas of volume-2 should be running on node-1, node-2, and node-3.

Addressed.

File: manager/integration/tests/test_zone.py [hunk 7]
```diff
@@ -104,11 +109,35 @@ def wait_longhorn_node_zone_updated(client): # NOQA
...
+        for j in range(RETRY_COUNTS):
+            lh_node = client.by_id_node(node_name)
+            if lh_node.zone == '':
+                break
+            time.sleep(RETRY_INTERVAL)
+
+        assert lh_node.zone == ''
+
+
+def reset_longhorn_node_zone(client, core_api):  # NOQA
```
[the remark this answers is not in the document]
Comment (c3y1huang) [reply manager/integration/tests/test_zone.py]: Thanks, addressed.

File: manager/integration/tests/test_zone.py (hunk 1 above)
Comment (c3y1huang) [reply manager/integration/tests/test_zone.py]: This is a typo, fixed to `<pod-node>`.

I am adapting to Gherkin Syntax, `< >` is the way presenting the scenario is running multiple times at different values. To specify the node name would duplicate the steps of the skeleton.

The values are matching
```
    Examples:        | pod-node | duplicate-node |        | node-1   | node-2         |        | node-2   | node-1         |        | node-1   | node-2         |
```

File: manager/integration/tests/test_zone.py (hunk 2 above)
Comment (c3y1huang) [reply manager/integration/tests/test_zone.py:536]: Already included the in the scenario with the loop through of deploying to `node-1` > `node-2` > `node-1`.

File: manager/integration/tests/test_scheduling.py [hunk 8]
```diff
@@ -422,6 +424,376 @@ def test_replica_rebuild_per_volume_limit(
...
+    And 6 replicas running on node-1.
+        0 replicas running on node-2.
+        0 replicas running on node-3.
+
+    When enable scheduling for node-2.
+    Then count replicas running on each nodes.
+    And node-1 replica count != node-2 replica count.
+        node-2 replica count != 0.
+        node-3 replica count == 0.
+    And sleep 10 seconds, to ensure no addition scheduling is happening.
```
[the remark this answers is not in the document]
Comment (c3y1huang) [reply manager/integration/tests/test_scheduling.py]: Addressed.

File: manager/integration/tests/test_scheduling.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (c3y1huang) [reply manager/integration/tests/test_scheduling.py]: Fixed.

**Comment by c3y1huang** [comment]
Regression * 3 - [Build#275](https://ci.longhorn.io/job/private/job/longhorn-tests-regression/275/testReport/tests/)

**Review by shuo-wu — APPROVED** [review]

**Pull request merged**